### PR TITLE
Use FSDP everywhere

### LINKED ISSUE #116: Deepspeed and bf16-true
In the finetuning scripts, we only allow

```python
precision: Literal["bf16-true", "32-true"] = "bf16-true",
```

But we also use DeepSpeed when `devices > 1`. However, in this case, you'd get a

```python
ValueError: `precision='bf16-true')` is not supported in DeepSpeed. `precision` must be one of: ('32-true', '16-mixed', 'bf16-mixed').
```

Should we allow `bf16-mixed`, or should we switch to FSDP? Or something else?

### LINKED ISSUE #169: Multiple GPU training stuck on Loading extension module utils / Tijme to load utils op XXX seconds
When starting the training using multiple GPU, it gets stuck indefinitely (I let it stuck for 10 minutes then I closed it) on the following log. Runpod.io tells me that the GPU have 34% VRAM used and 100% GPU Usage : 

With 2 gpu:
```
Loading extension module utils...
Time to load utils op: 11.247538805007935 seconds
Loading extension module utils...
Time to load utils op: 11.123851776123047 seconds
```

With 4 it looks like this:
```
Loading extension module utils...
Time to load utils op: 11.247538805007935 seconds
Loading extension module utils...
Time to load utils op: 11.123851776123047 seconds
Loading extension module utils...
Time to load utils op: 11.247538805007935 seconds
Loading extension module utils...
Time to load utils op: 11.123851776123047 seconds
```

Here is the stack trace :

```
root@492f6f0b65c4:/workspace/Lit-Parrot# ./start_training.sh 
[2023-06-19 15:41:32,009] [INFO] [real_accelerator.py:110:get_accelerator] Setting ds_accelerator to cuda (auto detect)
{'eval_interval': 8, 'save_interval': 20, 'eval_iters': 100, 'log_interval': 1, 'devices': 2, 'learning_rate': 0.0003, 'batch_size': 128, 'micro_batch_size': 2, 'gradient_accumulation_iters': 64, 'num_epochs': 60, 'max_iters': 31320, 'weight_decay': 0.01, 'lora_r': 8, 'lora_alpha': 16, 'lora_dropout': 0.05, 'warmup_iters': 100}
initializing deepspeed distributed: GLOBAL_RANK: 0, MEMBER: 1/2
[2023-06-19 15:41:32,209] [WARNING] [comm.py:152:init_deepspeed_backend] NCCL backend in DeepSpeed not yet implemented
[2023-06-19 15:41:35,207] [INFO] [real_accelerator.py:110:get_accelerator] Setting ds_accelerator to cuda (auto detect)
{'eval_interval': 8, 'save_interval': 20, 'eval_iters': 100, 'log_interval': 1, 'devices': 2, 'learning_rate': 0.0003, 'batch_size': 128, 'micro_batch_size': 2, 'gradient_accumulation_iters': 64, 'num_epochs': 60, 'max_iters': 31320, 'weight_decay': 0.01, 'lora_r': 8, 'lora_alpha': 16, 'lora_dropout': 0.05, 'warmup_iters': 100}
initializing deepspeed distributed: GLOBAL_RANK: 1, MEMBER: 2/2
[2023-06-19 15:41:35,376] [WARNING] [comm.py:152:init_deepspeed_backend] NCCL backend in DeepSpeed not yet implemented
Enabling DeepSpeed BF16.
[rank: 1] Global seed set to 1338
/usr/local/lib/python3.10/dist-packages/lightning/fabric/loggers/csv_logs.py:186: UserWarning: Experiment logs directory trained-model/falcon-7b-dsul_1/version_0 exists and is not empty. Previous log files in this directory will be deleted when the new ones are saved!
  rank_zero_warn(
[rank: 0] Global seed set to 1337
Loading model 'checkpoints/tiiuae/falcon-7b/lit_model.pth' with {'block_size': 2048, 'vocab_size': 50254, 'padding_multiple': 512, 'padded_vocab_size': 65024, 'n_layer': 32, 'n_head': 71, 'n_embd': 4544, 'rotary_percentage': 1.0, 'parallel_residual': True, 'bias': False, 'n_query_groups': 1, 'shared_attention_norm': True}
^[[CNumber of trainable parameters: 3506176
Using /root/.cache/torch_extensions/py310_cu118 as PyTorch extensions root...
Creating extension directory /root/.cache/torch_extensions/py310_cu118/utils...
Emitting ninja build file /root/.cache/torch_extensions/py310_cu118/utils/build.ninja...
Building extension module utils...
Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
Using /root/.cache/torch_extensions/py310_cu118 as PyTorch extensions root...
[1/2] c++ -MMD -MF flatten_unflatten.o.d -DTORCH_EXTENSION_NAME=utils -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1011\" -isystem /usr/local/lib/python3.10/dist-packages/torch/include -isystem /usr/local/lib/python3.10/dist-packages/torch/include/torch/csrc/api/include -isystem /usr/local/lib/python3.10/dist-packages/torch/include/TH -isystem /usr/local/lib/python3.10/dist-packages/torch/include/THC -isystem /usr/include/python3.10 -D_GLIBCXX_USE_CXX11_ABI=0 -fPIC -std=c++17 -c /usr/local/lib/python3.10/dist-packages/deepspeed/ops/csrc/utils/flatten_unflatten.cpp -o flatten_unflatten.o 
[2/2] c++ flatten_unflatten.o -shared -L/usr/local/lib/python3.10/dist-packages/torch/lib -lc10 -ltorch_cpu -ltorch -ltorch_python -o utils.so
Loading extension module utils...
Time to load utils op: 11.247538805007935 seconds
Loading extension module utils...
Time to load utils op: 11.123851776123047 seconds
```

I don't have this error using only one GPU.

### LINKED ISSUE #177: Cuda OOM : Falcon 7B with block size 3072
Hi,

This might not be possible yet but I am trying to fine-tune Falcon 7B with a block size of `3072` not `2048`.  I have two `49 GB A6000` and am trying to train with 1 device.  using `2048` takes up about 42GB of memory and then I get `OOM` with `3072`.  

I am using deepspeed stage 2 and would use stage 3 but it's not working because of the known issue.

Any ideas or should I wait for a stage 3 fix / quantization support?

Thanks,
Griffin

### PR DESCRIPTION
**Pull request by carmocca**
Replaces DeepSpeed with FSDP

Requires https://github.com/Lightning-AI/lightning/pull/17845

Closes #116 
Closes #177 
Closes #169

Falcon 7b takes 32 GB max memory allocated using 2 devices and 32-true or bf16-mixed precision.
Loss is NaN, requires further debugging into why.

### CODE REVIEW

File: lit_parrot/adapter.py [hunk 1]
```diff
@@ -254,6 +255,21 @@ def mark_only_adapter_as_trainable(model: Parrot) -> None:
         param.requires_grad = "adapter_wte" in name or "gating_factor" in name
 
 
-def adapter_state_from_state_dict(state_dict: dict) -> dict:
-    """Returns the model state dict with only the adapter weights for saving."""
-    return {name: param for name, param in state_dict.items() if "adapter_wte" in name or "gating_factor" in name}
+@contextlib.contextmanager
+def adapter_state_only(module: nn.Module):
```
Comment (carmocca) [inline lit_parrot/adapter.py:259]: FSDP didn't like this: https://github.com/pytorch/pytorch/issues/103136

File: finetune/adapter.py [hunk 2]
```diff
@@ -78,15 +74,14 @@ def main(
     mark_only_adapter_as_trainable(model)
 
     num_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
-    print(f"Number of trainable parameters: {num_params}")
+    fabric.print(f"Number of trainable parameters: {num_params}")
```
Comment (rasbt) [inline finetune/adapter.py:77]: Just curious what this does compared to regular `print`. In this case, if you are running on 8 GPUs, for example, the print statement would not be duplicated in the output terminal?

File: finetune/lora.py [hunk 3]
```diff
@@ -61,13 +55,16 @@ def setup(
...
-    # For multi-host TPU training, the device count for Fabric is limited to the count on a single host.
-    fabric_devices = "auto" if (tpu and devices > 1) else devices
+    fabric_devices = devices
+    if fabric_devices > 1:
+        if tpu:
+            # For multi-host TPU training, the device count for Fabric is limited to the count on a single host.
+            fabric_devices = "auto"
+            strategy = XLAStrategy(sync_module_states=False)
+        else:
+            raise NotImplementedError
```
Comment (carmocca) [inline finetune/lora.py:65]: The LoRA implementation doesn't work with FSDP:

```python
Traceback (most recent call last):
  File "finetune/lora.py", line 271, in <module>
    CLI(setup)
  File "/home/carlos/venv/lib/python3.8/site-packages/jsonargparse/cli.py", line 82, in CLI
    return _run_component(component, cfg_init)
  File "/home/carlos/venv/lib/python3.8/site-packages/jsonargparse/cli.py", line 138, in _run_component
    return component(**cfg)
  File "finetune/lora.py", line 79, in setup
    fabric.launch(main, data_dir, checkpoint_dir, out_dir, precision)
  File "/home/carlos/lightning/src/lightning/fabric/fabric.py", line 773, in launch
    return self._wrap_and_launch(function, self, *args, **kwargs)
  File "/home/carlos/lightning/src/lightning/fabric/fabric.py", line 854, in _wrap_and_launch
    return launcher.launch(to_run, *args, **kwargs)
  File "/home/carlos/lightning/src/lightning/fabric/strategies/launchers/subprocess_script.py", line 90, in launch
    return function(*args, **kwargs)
  File "/home/carlos/lightning/src/lightning/fabric/fabric.py", line 860, in _wrap_with_setup
Traceback (most recent call last):
  File "/home/carlos/lit-parrot/finetune/lora.py", line 271, in <module>
    return to_run(*args, **kwargs)
  File "finetune/lora.py", line 116, in main
    train(fabric, model, optimizer, train_data, val_data, checkpoint_dir, out_dir, max_seq_length, speed_monitor)
  File "finetune/lora.py", line 137, in train
    validate(fabric, model, val_data, tokenizer, max_seq_length)  # sanity check
  File "/home/carlos/venv/lib/python3.8/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    CLI(setup)
  File "/home/carlos/venv/lib/python3.8/site-packages/jsonargparse/cli.py", line 82, in CLI
    return func(*args, **kwargs)
    return _run_component(component, cfg_init)
  File "finetune/lora.py", line 215, in validate
  File "/home/carlos/venv/lib/python3.8/site-packages/jsonargparse/cli.py", line 138, in _run_component
    return component(**cfg)
    model.eval()
  File "/home/carlos/lit-parrot/finetune/lora.py", line 79, in setup
  File "/home/carlos/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 2321, in eval
    fabric.launch(main, data_dir, checkpoint_dir, out_dir, precision)
  File "/home/carlos/lightning/src/lightning/fabric/fabric.py", line 773, in launch
    return self._wrap_and_launch(function, self, *args, **kwargs)
  File "/home/carlos/lightning/src/lightning/fabric/fabric.py", line 855, in _wrap_and_launch
    return self.train(False)
  File "/home/carlos/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 2302, in train
    return to_run(*args, **kwargs)
  File "/home/carlos/lightning/src/lightning/fabric/fabric.py", line 860, in _wrap_with_setup
    return to_run(*args, **kwargs)
  File "/home/carlos/lit-parrot/finetune/lora.py", line 116, in main
    train(fabric, model, optimizer, train_data, val_data, checkpoint_dir, out_dir, max_seq_length, speed_monitor)
  File "/home/carlos/lit-parrot/finetune/lora.py", line 137, in train
    validate(fabric, model, val_data, tokenizer, max_seq_length)  # sanity check
  File "/home/carlos/venv/lib/python3.8/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/carlos/lit-parrot/finetune/lora.py", line 215, in validate
    module.train(mode)
  File "/home/carlos/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 2302, in train
    model.eval()
  File "/home/carlos/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 2321, in eval
    module.train(mode)
  File "/home/carlos/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 2302, in train
    return self.train(False)
  File "/home/carlos/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 2302, in train
    module.train(mode)
  [Previous line repeated 6 more times]
  File "/home/carlos/lit-parrot/lit_parrot/lora.py", line 271, in train
    module.train(mode)
  File "/home/carlos/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 2302, in train
    delta_w = F.conv1d(
RuntimeError: weight should have at least three dimensions
```

**Comment by MasterEndless** [comment]
I am using V100 GPU, and there is a bug to run the code:
KeyError: 'flop count not found for v100-sxm with precision: bf16-mixed; MFU cannot be calculated and reported.'

**Comment by carmocca** [comment]
@MasterEndless v100 cards do not support bf16 precision, try passing `--precision 32-true`

**Comment by MasterEndless** [comment]
@carmocca Thanks! I have tried, but seems a further error raises: 
 File "/home/ubuntu/disk2/lit-parrot/generate/base.py", line 75, in generate
    idx_next = torch.multinomial(probs, num_samples=1).to(dtype=dtype)
RuntimeError: probability tensor contains either `inf`, `nan` or element < 0
This happens in here: validate(fabric, model, val_data, tokenizer, max_seq_length)  # sanity check
Do you have any idea on why this bug happen? Thanks in advance.

**Comment by carmocca** [comment]
Yes, I noticed that issue, opened #185 to track it.

Which checkpoint are you using?

**Comment by MasterEndless** [comment]
@carmocca I am using falcon-7B-instruct

**Pull request merged**